addons/event_crm/models/event_registration.py

```diff
@@ -0,0 +1,274 @@
+            updating question answers or a contact information (email, phone);
+
+        Also trigger rules based on confirmed and done attendees (state written
+        to open and done).
+        """
+        to_update = self.filtered(lambda reg: reg.lead_ids)
+        if to_update:
+            lead_tracked_vals = to_update._get_lead_tracked_values()
+
+        res = super(EventRegistration, self).write(vals)
+
+        if to_update:
+            to_update.flush()  # compute notably partner-based fields if necessary
+            to_update.sudo()._update_lead_values(vals, lead_tracked_vals)
+
+        if vals.get('state') == 'open':
+            self.env['event.lead.rule'].search([('lead_creation_trigger', '=', 'confirm')]).sudo()._create_leads(self)
+        elif vals.get('state') == 'done':
+            self.env['event.lead.rule'].search([('lead_creation_trigger', '=', 'done')]).sudo()._create_leads(self)
+
+        return res
+
+    def _update_lead_values(self, new_vals, lead_tracked_vals):
+        """ Update leads linked to some registrations. Update is based depending
+        on updated fields, see ``_get_lead_contact_fields()`` and ``_get_lead_
+        description_fields()``. Main heuristic is
+
+          * check attendee-based leads, for each registration recompute contact
+            information if necessary (changing partner triggers the whole contact
+            computation); update description if necessary;
+          * check order-based leads, for each existing group-based lead, only
+            partner change triggers a contact and description update. We consider
+            that group-based rule works mainly with the main contact and less
+            with further details of registrations. Those can be found in stat
+            button if necessary.
+
+        :param new_vals: values given to write. Used to determine updated fields;
+        :param lead_tracked_vals: dict(registration_id, registration previous values)
+          based on new_vals;
+        """
+        for registration in self:
+            att_leads = registration.lead_ids.filtered(lambda lead: lead.event_lead_rule_id.lead_creation_basis == 'attendee')
+            if not att_leads:
+                continue
+
+            old_vals = lead_tracked_vals[registration.id]
+            # if partner has been touched -> update registration contact information as they are computed
+            if 'partner_id' in new_vals:
+                new_vals.update(**dict((field, registration[field]) for field in ('name', 'email', 'phone', 'mobile')))
+
+            lead_values = {}
+            if any(new_vals[field] != old_vals[field] for field in self._get_lead_contact_fields() if field in new_vals.keys()):
+                lead_values = registration._get_lead_contact_values()
+
+            for lead in att_leads:
+                if any(new_vals[field] != old_vals[field] for field in self._get_lead_description_fields() if field in new_vals.keys()):
```

Comment:
Indeed, triggers are not completely the same. I was a bit lost myself, but this is the spec: description holds registration contact + answers (basically), contact is registration contact + partner (because managed a bit differently from other contact field). Tracked vals is "all interesting vals for our update".

Reply:
Added a comment.

Reply:
I mean, is it possible that some of the contact_fields are not present in the tracked field ? if yes, I guess we will have a problem getting `tracked[field] for field in contact_fields`

Reply:
Normally not, as we check for fields in new_vals (updated), and those are used to build tracked. Did you find a specific issue or are you just wondering ?